### ISSUE TITLE
Enable to customize ExecutionLoop Listeners

### ISSUE THREAD

**Issue by IonBazan** [bobthecow/psysh#565]
Currently it is no easy way to customize the execution loop listeners. 
I managed to do that by extending `Psy\Shell` and overriding `getDefaultLoopListeners()` method:
```php
namespace App\Console\Shell;

use Psy\Shell;

class MyShell extends Shell
{
    protected function getDefaultLoopListeners()
    {
        $listeners = [new MyLoopListener()];

        return array_merge($listeners, parent::getDefaultLoopListeners());
    }
}
```
but that doesn't seem like a clean solution. I think we should enable users to customize them via constructor or `addLoopListener()`/`setLoopListeners()` methods following `addMatchers()` and  `addTabCompletionMatchers()` methods convention so the code would become:
```php
$shell = new Shell();
$shell->addLoopListener(new MyLoopListener());
```

@bobthecow WDYT?

**Comment by bobthecow** [comment]
I understand _what_ you want, but can you share a bit more about _why_? 🙂

**Comment by IonBazan** [comment]
I was trying to log every `beforeRun`, `onExecute` and `afterRun` event. I ended up creating a class inheriting the `Shell` and overriding these methods to add such behavior but it is not that flexible (composition over inheritance).

**Comment by bobthecow** [comment]
Right, but why? I still don’t understand your use case. I’m not opposed to opening up more extension points, but there _is_ a maintenance and support cost associated, and this one hasn’t happened because I haven’t seen a good reason for it yet. 

>As a ___________ I want to ___________ so that ___________

🙂

**Comment by bobthecow** [comment]
Logging is available in the latest release!
